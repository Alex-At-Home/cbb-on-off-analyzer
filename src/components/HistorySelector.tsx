// React imports:
import React, { useState, useEffect } from 'react';

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';

// Additional components:
import Select, { components} from "react-select"

// Utils:
import { HistoryManager } from "../utils/HistoryManager"

/** Popovers appear to need to know their children's width, if it's "complex" */
export const historySelectContainerWidth = "600px";

const HistorySelector: React.FunctionComponent<{}> = ({}) => {

  /** For use in selects */
  function stringToOption(s: string) {
    return { label: s, value: s};
  }
  function getPlaceholder() {
    return { label: 'Select analysis from history...' };
  }

  return <div style={{ width: "500px" }}>
    <Select
      isClearable={true}
      styles={{ menu: base => ({ ...base, zIndex: 1000 }) }}
      value={ getPlaceholder() }
      options={HistoryManager.returnHistory().map(
        (r) => stringToOption("TODO")
      )}
      onChange={(option) => {
        //TODO
      }}
    />
  </div>;
}
export default HistorySelector;
