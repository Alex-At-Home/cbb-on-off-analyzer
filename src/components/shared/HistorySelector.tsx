// React imports:
import React, { useState, useEffect } from "react";
import Router from "next/router";

// Additional components:
// @ts-ignore
import Select, { components } from "react-select";

// Utils:
import { QueryUtils } from "../../utils/QueryUtils";
import { UrlRouting } from "../../utils/UrlRouting";
import { HistoryManager } from "../../utils/HistoryManager";
import {
  ParamPrefixes,
  GameFilterParams,
  LineupFilterParams,
  TeamReportFilterParams,
} from "../../utils/FilterModels";

/** Popovers appear to need to know their children's width, if it's "complex" */
export const historySelectContainerWidth = "600px";

type Props = {
  tablePrefix: string;
};
const HistorySelector: React.FunctionComponent<Props> = ({ tablePrefix }) => {
  const allParams =
    typeof window === `undefined` //(ensures SSR code still compiles)
      ? ""
      : window.location.search;

  /** For use in selects */
  function stringToOption(
    s: string,
    val: [string, GameFilterParams | LineupFilterParams]
  ) {
    return { label: s, value: JSON.stringify(val) };
  }
  function getPlaceholder() {
    return { label: "Select analysis from history..." };
  }

  return (
    <div style={{ width: "500px" }}>
      <Select
        styles={{ menu: (base: any) => ({ ...base, zIndex: 1000 }) }}
        value={getPlaceholder()}
        menuIsOpen={true}
        options={HistoryManager.getHistory().map((r) =>
          stringToOption(HistoryManager.filterSummary(r[0], r[1]), r)
        )}
        onChange={(option: any) => {
          const valJson = JSON.parse((option as any)?.value || "{}"); // [string, GameFilterParams | LineupFilterParams]

          const getUrl = () => {
            if (valJson[0] == ParamPrefixes.game) {
              return UrlRouting.getGameUrl(valJson[1] as GameFilterParams, {});
            } else if (valJson[0] == ParamPrefixes.lineup) {
              return UrlRouting.getLineupUrl(
                valJson[1] as LineupFilterParams,
                {}
              );
            } else if (valJson[0] == ParamPrefixes.report) {
              return UrlRouting.getTeamReportUrl(
                valJson[1] as TeamReportFilterParams
              );
            } else {
              return undefined;
            }
          };
          const newUrl = getUrl();
          if (newUrl) {
            if (tablePrefix == valJson[0]) {
              // (can't force a full client refresh using Router - this is an ugly alternative)
              window.location.href = newUrl;
            } else {
              Router.push(newUrl);
            }
          } else {
            //(else do nothing)
            console.log(
              `[WARNING] unexpected history selection: ${JSON.stringify(
                option
              )}`
            );
          }
        }}
      />
    </div>
  );
};
export default HistorySelector;
