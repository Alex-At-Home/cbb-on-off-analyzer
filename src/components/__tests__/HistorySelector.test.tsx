import React from 'react';
import HistorySelector from '../HistorySelector';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe("HistorySelector", () => {
  test("HistorySelector - should create snapshot", () => {
    const wrapper = shallow(<HistorySelector
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
