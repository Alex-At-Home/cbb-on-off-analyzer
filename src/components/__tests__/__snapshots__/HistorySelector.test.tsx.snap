// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistorySelector HistorySelector - should create snapshot 1`] = `
<div
  style={
    Object {
      "width": "500px",
    }
  }
>
  <StateManager
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isClearable={true}
    onChange={[Function]}
    options={Array []}
    styles={
      Object {
        "menu": [Function],
      }
    }
    value={
      Object {
        "label": "Select analysis from history...",
      }
    }
  />
</div>
`;
